tensordict/nn/common.py

```diff
@@ -25,14 +24,9 @@
     _has_functorch = False
 
 import torch
-from functorch import FunctionalModule, FunctionalModuleWithBuffers, vmap
-from functorch._src.make_functional import _swap_state
+from functorch import FunctionalModule, FunctionalModuleWithBuffers
```

Comment:
Looking at the API, nn.utils.stateless.functional_call seems to provide something similar (in-place functional modules, but not exactly):

```
l = nn.Sequential(
    Linear(3, 4),
    LinearWithKwargs(4, 4),
)
params = l.named_parameters()
x = torch.randn(3)
nn.utils.stateless.functional_call(l, params, x)
```
What are the advantages of this tensordict-based approach?

> Another important deviation to note is that we're just using Params, not buffers.
This is highly opinionated but I don't really see the point of having 2 different pipelines, one for params and one for params and buffers.

I never liked the old make_functional design where we separated modules that had params and modules had params+buffers

Reply:
> What are the advantages of this tensordict-based approach?

I don't know if everyone would agree about the fact that these are advantages, but here's my take:
- You can access the params/buffers in a very similar way as you access your modules. For instance, once you've called `make_functional` you can do `module.sub_module[some_index](input, params["sub_module", "some_index"]`. Think of a transformer where you need to query the encoder / decoder separately. You can collect your params in a TensorDict just once and play with it as you like. With the pytorch approach, you'd need either (1) to query `named_parameters` for each sub-module (which can be expensive) or (2) write some complicated indexing methods that, actually, are taken care of by tensordict.
-  You benefit from all the tensordict features (compat with vmap, indexing, reshaping, stacking for ensembling, pre-allocation etc)
Here are some examples: 
```python
vmap(module.sub_module, (None, 0))(x, params["sub_module"].expand(10))
module.sub_module(x.cuda(), params["sub_module"].cuda())
module(x, params.detach())
```
- You query your model directly, no need for a helper function. IMO it's much more good-looking, and you can revert things and get back your original module. With `functional_call` you still have a model that is full of params, which may come with some problems (ie. memory allocation of a whole network, or loosing sight of why those params are there etc).
- You have a functional module where you can replace all the params if you want using `_swap_state` (though I guess the torch implementation does something similar under the hood).
Hope that clarifies things a bit!